packages/devtools/lib/src/timeline/flame_chart_canvas.dart

```diff
@@ -0,0 +1,624 @@
+    );
+
+    String displayText = text.substring(0, maxLetters);
+    if (maxLetters == maxDisplayTextLength &&
+        text.length > maxDisplayTextLength) {
+      displayText += '...';
+    }
+
+    canvas.fillText(
+      displayText,
+      rect.left + horizontalPadding,
+      rect.top + textOffsetY,
+    );
+  }
+
+  void updateHorizontalPosition({@required num zoom}) {
+    // Do not round these values. Rounding the left could cause us to have
+    // inaccurately placed events on the chart. Rounding the width could cause
+    // us to lose very small events if the width rounds to zero.
+    final newLeft = (startingLeft - _flameChartInset) * zoom + _flameChartInset;
+    final newWidth = startingWidth * zoom;
+
+    final updatedRect = Rect.fromLTWH(newLeft, rect.top, newWidth, rect.height);
+    rect = updatedRect;
+  }
+}
+
+class TimelineGrid {
+  TimelineGrid(this._duration, this._flameChartWidth);
+
+  static const baseGridInterval = 150;
+
+  /// Frame duration in micros.
+  final num _duration;
+
+  num currentInterval = baseGridInterval;
+
+  num _flameChartWidth;
+
+  num _zoomLevel = 1;
+
+  void paint(CanvasRenderingContext2D canvas, Rect viewport, Rect visible) {
+    // Draw the first grid item since it will have a different width than the
+    // rest.
+    final gridItem = TimelineGridNode(
+      0,
+      _flameChartInset,
+      const Duration(microseconds: 0),
+    );
+
+    gridItem.paint(canvas, viewport, visible);
+
+    num left =
+        (visible.left - _flameChartInset) ~/ currentInterval * currentInterval +
+            _flameChartInset;
+
+    while (left < visible.right) {
+      // TODO(kenzie): Instead of calculating timestamp based on position, track
+      // timestamp var and increment it by time interval represented by each
+      // grid item. See comment on https://github.com/flutter/devtools/pull/325.
+      final Duration timestamp = Duration(
+          microseconds: getTimestampForPosition(left + currentInterval));
+      final gridItem = TimelineGridNode(left, currentInterval, timestamp);
+
+      gridItem.paint(canvas, viewport, visible);
+
+      left += currentInterval;
+    }
+  }
+
+  /// Returns the timestamp rounded to the nearest microsecond for the
+  /// x-position.
+  int getTimestampForPosition(num gridItemEnd) {
+    return ((gridItemEnd - _flameChartInset) / _flameChartWidth * _duration)
+        .round();
+  }
+
+  void updateForZoom(num newZoomLevel, num newFlameChartWidth) {
+    if (_zoomLevel == newZoomLevel) {
+      return;
+    }
+
+    _flameChartWidth = newFlameChartWidth;
+
+    final log2NewZoomLevel = log2(newZoomLevel);
+
+    final gridZoomFactor = math.pow(2, log2NewZoomLevel);
+    final gridIntervalPx = baseGridInterval / gridZoomFactor;
+
+    /// The physical pixel width of the grid interval at [newZoomLevel].
+    currentInterval = gridIntervalPx * newZoomLevel;
+
+    _zoomLevel = newZoomLevel;
+  }
+}
+
+class TimelineGridNode {
```

Comment:
add a TODO to delete this class as it really isn't needed but keeping it for now is fine.

Reply:
done.